pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go

```diff
@@ -82,4 +84,38 @@ var _ = Describe("Qemu agent poller", func() {
 			Expect(agentStore.AgentUpdated).ToNot(Receive())
 		})
 	})
+
+	Context("PollerWorker", func() {
+		const fakeAgentCommandName = "foo"
+
+		It("executes the agent commands at least once", func() {
+			w := PollerWorker{
+				CallTick:      1,
+				AgentCommands: []AgentCommand{fakeAgentCommandName},
+			}
+			c := make(chan struct{})
+
+			var execCounter int
+			go w.Poll(func(commands []AgentCommand) { execCounter++ }, c)
+			close(c)
+
+			Eventually(func() int { return execCounter }, 5*time.Millisecond, time.Millisecond).Should(Equal(1))
+		})
+
+		It("executes the agent commands based on the time interval specified", func() {
+			w := PollerWorker{
+				CallTick:      1,
+				AgentCommands: []AgentCommand{fakeAgentCommandName},
+			}
+			c := make(chan struct{})
+
+			var execCounter int
+			go w.Poll(func(commands []AgentCommand) { execCounter++ }, c)
+			time.Sleep(2*time.Second + time.Millisecond)
```

Comment:
here i wont trust it, it might flake, 
you can have context switch + page fault and the go routine wont manage to do the job
i would sleep much more than that, or use Eventually instead sleep

whole story at
https://github.com/kubevirt/kubevirt/pull/4512/commits/a2fbf7e0b67a259040ea54bec4a978e5bb1e1737#r521139630

Reply:
- It has a larger sleep (2sec + 1msec) to handle such glitches.
- There is an eventually in place to wait for the goroutine to finish.

Any context switch in the groutine will not affect the time of the select at a noticeable scale, it uses channel signals inside for the tick. For the overhead, the msec is used.

You are welcome to provide a full flow of the scenario you see as problematic or reference the flake.
At the moment, I do not see the failure point.

Reply:
There can be a race between:
1. Close of the channel
2. go routine managing to consume the ticket and do the job

if 1 happens before N times that 2 should happen it would flake.

Reply:
> if 1 happens before N times that 2 should happen it would flake,
> because the select might choose the channel close event if both happened.

I do not understand this, please rephrase. Actually, better give a step by step scenario where there is a problem.
For example:
- t0 goroutine is called.
- t1 sleep starts.
- t2 goroutine starts, executes command, waits for tick.
- t3 channel closes.
...

Reply:
Lets call the main code that runs line 112 and so on `g1`
Let `w.Poll` go routine be 'g2`

* t0 g2 is called.
* t1 sleep starts.
* t2 - t4 g2 executes command, consume first tick (now counter is 2),  waits for the 2nd tick .
* t3 g1 sleep finish , and close channel.
* t4 g2 didnt mange to consume the ticker before the channel is closed, it consume the channel close before
(1 millisecond isnt a glitch proof in other words)

Reply:
The tick is called every second, so you need to explain what can happen that the 2nd tick will not arrive after ~2sec.
The 1msec is the inaccuracy of the ticker.

So what will make the ticker not to return on the second time after 2 sec + some random drift of up to 1msec?

Reply:
> The tick is called every second, so you need to explain what can happen that the 2nd tick will not arrive after ~2sec.
> The 1msec is the inaccuracy of the ticker.
> 
> So what will make the ticker not to return on the second time after 2 sec + some random drift of up to 1msec?

Its not the ticket wont return, its the `g2` that wont get scheduled in time,
various reasons can cause it, interrupts, page faults, syscalls that move the context to the kernel,
the test is a process in a whole OS.

Reply:
So you claim that the go runtime will not run the goroutine when the tick will arrive? And that will take more than 1msec?
1msec is a big offset and the fact that a channel signal in playing a part here (will trigger the runtime to run the goroutine) I do not see how this can ever happen.

Lets assume this can happen, what is your solution?

Reply:
It can happen, but lets continue according your last line.

Possible solutions:
1
a. Use Eventually that the counter need to reach count N.
b. close channel
c. Use Consistently count should reach a value and stop incrementing
2. 
a. use bigger sleep
b. check the timer stop changing, and that its value is >= 3

Reply:
> 1. Use eventually to "close" in the correct time

This is possible, but then if there is indeed a bug and we never reach the relevant value it will block with no timeout.
Not really sure why there is a need to check consistently.

> 2. Bigger sleep and check for >=3

1msec sleep is a lot for a computer, I did not use nanoseconds. But I am fine to increase it to 10msec if that will satisfy your concerns.


Checking that the value is not changing anymore is testing the "close" operation, but I feel it is redundant, especially with unit tests.
I will wait for a second opinion in this one, seems borderline to me.

Reply:
> > 1. Use eventually to "close" in the correct time
> 
> This is possible, but then if there is indeed a bug and we never reach the relevant value it will block with no timeout.
> Not really sure why there is a need to check consistently.

You will use a timeout Eventually, all will be backed up with timeout, nothing infinite.

> 
> > 1. Bigger sleep and check for >=3
> 
> 1msec sleep is a lot for a computer, I did not use nanoseconds. But I am fine to increase it to 10msec if that will satisfy your concerns.
> 
> Checking that the value is not changing anymore is testing the "close" operation, but I feel it is redundant, especially with unit tests.
> I will wait for a second opinion in this one, seems borderline to me.

It has inherit problem which is adding a jiffy which might flake imo.

Reply:
> > > 1. Use eventually to "close" in the correct time
> > 
> > 
> > This is possible, but then if there is indeed a bug and we never reach the relevant value it will block with no timeout.
> > Not really sure why there is a need to check consistently.
> > > 1. Bigger sleep and check for >=3
> > 
> > 
> > 1msec sleep is a lot for a computer, I did not use nanoseconds. But I am fine to increase it to 10msec if that will satisfy your concerns.
> 
> +1
> 
> > Checking that the value is not changing anymore is testing the "close" operation, but I feel it is redundant, especially with unit tests.
> > I will wait for a second opinion in this one, seems borderline to me.
> 
> Do we want to test the close operation ? I.e. did we test it before ? I assume no.

Its so the flow would be complete, 
lets think we "simplify" it:
1. Eventually get 3
2. close
Done

For me it looks incomplete, as the new test indeed want to check an finite interval which is bounded.
Up to your decision, i would at least use more than 1 msec imo.

> 
> **EDIT**: having said / asked that, I'm forced to agree that `time.Sleep` smells weird... But since you're testing something based on a ticker, I personally don't find it problematic.

Reply:
I have changed the mechanism in the tests and removed the `sleep`.
Perhaps @oshoval you will like it better now.